File: src/main/java/org/radarbase/appserver/event/state/NotificationStateEvent.java
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.appserver.event.state;

import lombok.Getter;
import lombok.ToString;
import org.radarbase.appserver.entity.Notification;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;

import java.time.Instant;
import java.util.Map;

@Getter
@ToString
public class NotificationStateEvent extends MessageStateEvent {
    private static final long serialVersionUID = 327842183571938L;

    private Notification notification;

    /**
     * Create a new ApplicationEvent.
     *
     * @param source         the object on which the event initially occurred (never {@code null})
     * @param notification   the notification associated with this state event.
     * @param state          the current {@link MessageState} change of the {@link
     *                       Notification} entity.
     * @param additionalInfo any additional info associated with the state change.
     */
    public NotificationStateEvent(
            Object source,
            @NonNull Notification notification,
            @NonNull MessageState state,
            @Nullable Map<String, String> additionalInfo,
            @NonNull Instant time) {
        super(source, state, additionalInfo, time);
        this.notification = notification;
    }
}

File: appserver-jersey/src/main/kotlin/org/radarbase/appserver/jersey/entity/NotificationStateEvent.kt
/*
 * Copyright 2025 King's College London
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.appserver.jersey.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import jakarta.validation.constraints.NotNull
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import org.radarbase.appserver.jersey.event.state.MessageState
import java.time.Instant

@Entity
@Table(name = "notification_state_events")
class NotificationStateEvent : MessageStateEvent {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    var id: Long? = null

    @field:NotNull
    @field:JsonIgnore
    @ManyToOne(fetch = FetchType.EAGER, optional = false)
    @JoinColumn(name = "notification_id", nullable = false)
    @field:OnDelete(action = OnDeleteAction.CASCADE)
    var notification: Notification? = null

    constructor(
        notification: Notification,
        state: MessageState,
        time: Instant,
        associatedInfo: String?,
    ) : super(state, time, associatedInfo) {
        this.notification = notification
    }

    constructor()
}
